### PR TITLE
Request Latency Percentiles

### PR DESCRIPTION
**Pull request by mkralka**
- Introduce request latency percentiles
- Cherry-pick a performance improvement from master

### CODE REVIEW

File: memcached.c [hunk 1]
```diff
@@ -3279,7 +3356,7 @@ static void process_command(conn *c, char *command) {
     c->iovused = 0;
     if (add_msghdr(c) != 0) {
         out_string(c, "SERVER_ERROR out of memory preparing response");
-        return;
+        goto out;
```
Comment (kevinballard) [inline memcached.c]: This doesn't feel like a good idea...

File: memcached.c [hunk 2]
```diff
@@ -166,7 +168,16 @@ static rel_time_t realtime(const time_t exptime) {
...
+static void stats_report_latency(stat_cmd_t cmd, hires_time_t latency) {
+    if (cmd < STAT_CMD_COUNT) {
+        STATS_LOCK();
+        percentile_sampler_record_sample(&stats.per_cmd_stats[cmd].handle_latency, latency);
+        STATS_UNLOCK();
+    }
+}
+
 static void stats_init(void) {
+    size_t i;
```
Comment (kevinballard) [inline memcached.c:180]: Why size_t? It's counting indexes, so I would assume int or short would be more appropriate.

File: memcached.c [hunk 3]
```diff
@@ -187,6 +201,7 @@ static void stats_init(void) {
 }
 
 static void stats_reset(void) {
+    size_t i;
```
Comment (kevinballard) [inline memcached.c:204]: ditto re: size_t

File: memcached.c (hunk 1 above)
Comment (mkralka) [inline memcached.c]: What about this bothers you?

File: memcached.c (hunk 2 above)
Comment (mkralka) [inline memcached.c:180]: `i` is being compared against `sizeof(stats.per_cmd_stats)/sizeof(*stats.per_cmd_stats)` which is a `size_t`. This ensures no platform-dependent compiler warnings with differences between `size_t` and `int` (or whatever integral type is selected), such as signedness or bit-width.

File: memcached.c (hunk 1 above)
Comment (kevinballard) [inline memcached.c]: goto makes me nervous. Can we instead wrap the logic in its own function?

File: memcached.c (hunk 2 above)
Comment (kevinballard) [inline memcached.c:180]: Ok, makes sense.

File: memcached.c (hunk 1 above)
Comment (mkralka) [inline memcached.c]: Hrm...

I think a separate function is more prone to error as it requires diligence in being consistent with the call point for every failure case. As things stand, we default to "an error occurred" and clear than if the command succeeds, so we only need to keep that set properly on success paths.

Let me see if the code can be re-organized without  gotos without being too messy.

Comment (mkralka) [inline memcached.c]: I think we're probably OK with just `return` in all of these cases.

**Comment by kevinballard** [comment]
LGTM

**Pull request merged**